Allow filters to use extraction functions

jon-wei opened the pull request:
This PR allows the user to define an extraction function on each of the filters (except SpatialFilter); the extraction function will be applied to the input before the filter criteria is applied.

This merges the SelectorFilter and ExtractionFilter into a single class (ExtractionDimFilter is deprecated in this patch, ExtractionFilter has been deleted).

---

Fixes #2643

Also fixes #2772, #2775, #2776

jon-wei commented:
updated docs

jon-wei commented:
rebased, removed conflicts with https://github.com/druid-io/druid/pull/2711

navis commented:
I think `JavaScriptFilter` and `SelectorFilter` also can extend `DimensionPredicateFilter`, removing some duplicated codes.

navis commented on processing/src/main/java/io/druid/segment/filter/DimensionPredicateFilter.java:
```diff
@@ -37,14 +38,17 @@
       String dimension,
-      Predicate<String> predicate
+      Predicate<String> predicate,
+      ExtractionFn extractionFn
```
cause extraction itself is handled in predicate, I think we don't need this here (can be confusing).

jon-wei replied:
@navis thanks, extractionFn was unnecessary here, removed it

gianm replied:
but now it's necessary again, right?

jon-wei replied:
yes, it's being used now

navis commented on processing/src/main/java/io/druid/segment/filter/SelectorFilter.java:
```diff
@@ -31,25 +39,93 @@
+
+  private List<Filter> makeFilters(BitmapIndexSelector selector)
+  {
+    Indexed<String> allDimVals = selector.getDimensionValues(dimension);
```
If we use `Iterable` here instead of `Indexed`, null value for `allDimVals` can just be `Arrays.asList((String)null)` not making strange `Indexed`.

jon-wei replied:
@navis The code was copied from ExtractionFilter, but agreed, it looks cleaner with Iterable, changed it to that

gianm commented:
sorry @jon-wei, more conflicts!

jon-wei commented:
@navis 

re: changing SelectorFilter and JavascriptFilter to implement DimensionPredicateFilter

There are some performance considerations if that change were made:

- getBitmapIndex() for DimensionPredicateFilter requires a scan through the column's dictionary values to determine which values matched the predicate. For SelectorFilter without an extraction function, getBitmapIndex() is much simpler/faster since it can just call selector.getBitmapIndex(dimension, value).

- For JavascriptFilter, the current implementation allows getBitmapIndex() to reuse the Context object across the scan of the dictionary. Using the JavascriptPredicate with DimensionPredicateFilter's getBitmapIndex() function would result in a new Context object being created on every apply() call. There's a note regarding this performance consideration at:
https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/segment/filter/JavaScriptFilter.java#L92

I think, if that change is made, it's better handled in a separate PR.

jon-wei commented:
@gianm np, rebasing now

jon-wei commented:
rebased

navis commented:
@jon-wei Pair enough. :+1:

gianm commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -53,6 +56,19 @@ public BoundDimFilter(
+    this.extractionFn = extractionFn;
+  }
+
+  public BoundDimFilter(
```
Is this constructor needed? If it's only used for tests then usually we modify the tests rather than add another constructor.

(similar comment for all the other filters)

gianm replied:
IMO one nice thing about only having one constructor is that it makes it harder to accidentally create an object with a critical parameter missing.

jon-wei replied:
@gianm removed extra constructors

drcrallen commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -146,40 +173,44 @@ public boolean equals(Object o)
     BoundDimFilter that = (BoundDimFilter) o;
 
-    if (isLowerStrict() != that.isLowerStrict()) {
+    if (lowerStrict != that.lowerStrict) {
```
Is there a need to move from getters to fields here?

jon-wei replied:
@drcrallen regenerated equals() with getter methods

drcrallen commented on processing/src/main/java/io/druid/query/filter/ExtractionDimFilter.java:
```diff
@@ -29,12 +29,14 @@
 
 /**
  */
+@Deprecated
```
Add javadoc to replacement plz

jon-wei replied:
@drcrallen added link to SelectorDimFilter

gianm commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -108,11 +130,14 @@ public boolean isAlphaNumeric()
     byte upperStrictByte = (this.isUpperStrict() == false) ? 0x0 : (byte) 1;
     byte AlphaNumericByte = (this.isAlphaNumeric() == false) ? 0x0 : (byte) 1;
 
+    byte[] extractionFnBytes = extractionFn.getCacheKey();
```
`extractionFn` might be null

jon-wei replied:
@gianm changed that line to have a null check on extractionFn like the rest of the filters, added some cases with extractionFns to the filter cache key tests (e.g., BoundDimFilterTests)

gianm commented on docs/content/querying/filters.md:
````diff
@@ -81,35 +81,6 @@ The following matches any dimension values for the dimension `name` between `'ba
 }
 ```
 
-### Extraction filter
````
suggest leaving the docs in, but marking it deprecated/unnecessary.

jon-wei replied:
@gianm put the extraction filter docs back, marked it deprecated with a "note caution" block

drcrallen commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -96,27 +144,22 @@ public boolean equals(Object o)
     if (value != null ? !value.equals(that.value) : that.value != null) {
       return false;
     }
+    return extractionFn != null ? extractionFn.equals(that.extractionFn) : that.extractionFn == null;
```
Do all the extraction functions have proper equals implementations? I haven't actually checked.

drcrallen replied:
or what is the expected ramification if it does not?

jon-wei replied:
@drcrallen Only BoundDimFilter and SelectorDimFilter have proper equals implementations;

I had the equals() and hashcode() methods for those DimFilters throw an exception when called, and found out that AndDimFilter and OrDimFilter trigger equals() on its sub filters via this call:
```
    fields.removeAll(Collections.singletonList(null));
```

But the default equals() implementation is sufficient for this null removal.

I also disabled the proper equals/hashcode methods on BoundDimFilter and SelectorDimFilter, the only problem I saw with doing that in the unit tests was in TimeseriesQueryRunnerTest, where there are some explicit equality checks on SelectorDimFilter:
```
    Assert.assertEquals(
        Druids.newSelectorDimFilterBuilder()
              .dimension(QueryRunnerTestHelper.marketDimension)
              .value("upfront")
              .build(),
        query.getDimensionsFilter()
    );
```

gianm commented on docs/content/querying/filters.md:
```diff
@@ -270,3 +241,38 @@ Now suppose that the lower bound is `110` the filter will not match (`110 < 10K`
 |caseSensitive|Whether two string should be compared as case sensitive or not|yes|
+
+
+### Filtering with Extraction Functions
```
suggest also adding an `extractionFn` option to each of the individual filter docs, that links here.

jon-wei replied:
@gianm Added a link in each individual filter section pointing to "Filtering with Extraction Functions"

drcrallen commented on processing/src/main/java/io/druid/segment/filter/DimensionPredicateFilter.java:
```diff
@@ -23,6 +23,7 @@
 import com.google.common.base.Predicate;
 import com.metamx.collections.bitmap.ImmutableBitmap;
 import com.metamx.common.guava.FunctionalIterable;
+import io.druid.query.extraction.ExtractionFn;
```
unused?

jon-wei replied:
@drcrallen changed this per @gianm's suggestion to have the extractionFn applied in the DimensionPredicateFilter

drcrallen commented on processing/src/main/java/io/druid/segment/filter/ExtractionFilter.java:
```diff
@@ -35,6 +35,7 @@
 
 /**
  */
+@Deprecated
```
Add javadoc on replacement plz

jon-wei replied:
@drcrallen Added link to SelectorFilter

drcrallen commented on processing/src/main/java/io/druid/segment/filter/SelectorFilter.java:
```diff
@@ -31,25 +40,68 @@
+      if (filters.isEmpty()) {
+        return selector.getBitmapFactory().makeEmptyImmutableBitmap();
+      }
+      return new OrFilter(makeFilters(selector)).getBitmapIndex(selector);
```
does makeFilters need to be called again here? can you just use `filters` from previously?

jon-wei replied:
@drcrallen changed this to reuse ```filters```

gianm commented on processing/src/main/java/io/druid/segment/filter/BoundFilter.java:
```diff
@@ -59,6 +61,11 @@ public boolean apply(String input)
                     return false;
                   }
+
+                  if (extractionFn != null) {
```
I think it'd make more sense for DimensionPredicateFilter to take the `extractionFn` and apply it.

gianm replied:
This way the predicate supplied to DimensionPredicateFilter is something that operates on extracted values (which seems to make more sense to me)

jon-wei replied:
@gianm Reimplemented to apply extractionFn in DimensionPredicateFilter

gianm commented on processing/src/main/java/io/druid/segment/filter/JavaScriptFilter.java:
```diff
@@ -127,11 +132,13 @@ public boolean apply(final String input)
     public boolean applyInContext(Context cx, String input)
     {
+      if (extractionFn != null) {
+        input = extractionFn.apply(input);
```
`Strings.nullToEmpty(extractionFn.apply(input))`

gianm replied:
nevermind, `null` is ok.

gianm commented on processing/src/main/java/io/druid/segment/filter/BoundFilter.java:
```diff
@@ -59,6 +61,11 @@ public boolean apply(String input)
                   }
+
+                  if (extractionFn != null) {
+                    input = extractionFn.apply(input);
```
`Strings.nullToEmpty(extractionFn.apply(input))` (although I still think this would make more sense in DPF)

gianm replied:
actually nevermind. the comparators can handle null property- don't do `nullToEmpty`.

gianm commented on processing/src/main/java/io/druid/segment/filter/RegexFilter.java:
```diff
@@ -44,6 +47,17 @@ public boolean apply(String input)
+          @Override
+          public boolean apply(String input)
+          {
+            return (input != null) && compiled.matcher(extractionFn.apply(input)).find();
```
`Strings.nullToEmpty(extractionFn.apply(input))`

gianm replied:
nevermind -- this should be moved to DPF.

jon-wei replied:
@gianm moved to DPF

drcrallen commented on processing/src/main/java/io/druid/segment/filter/SelectorFilter.java:
```diff
@@ -31,25 +40,68 @@
+            public boolean apply(String input)
+            {
+              // Assuming that a null/absent/empty dimension are equivalent from the druid perspective
+              return value.equals(Strings.nullToEmpty(extractionFn.apply(Strings.emptyToNull(input))));
```
This feels like a lot of hoops to jump through.

The ExtractionFn contract probably needs clarified on what goes in. Just looking at a few spot cases it is not obvious things like the JavaScript extraction fn handles null inputs in a proper way.

Given that the output is always null and never empty (by method contract) it seems like we should be able to eliminate the nullToEmpty... and if the ability for ExtractionFn.apply to accept either null xor empty in the method contract, that should help clarify the need for the emptyToNull.

Some ExtractionFuncions seem to handle null input explicitly, and others do not seem to do so, which may be either a waiting bug or an unneeded check.

Anyways, such a change would be outside the scope of this PR and the approach here is fine for now, but please make sure `value` has a null check somewhere before it reaches here.

jon-wei replied:
@drcrallen I added a ```this.value = Strings.nullToEmpty(value);``` call to SelectorFilter's constructor

gianm replied:
I don't think the `Strings.emptyToNull` is needed on the input to the extractionFn; `input` shouldn't be empty anyway (ValueMatcherFactories aren't passing in empty strings to the predicate) and also most other callers of extractionFns don't do this.

drcrallen commented:
The approach here looks good.

I would like to see more unit tests covering the new code.

There are some field additions which are added to hashCode and equals, and some which are not. I assume the later is because the classes lack hachCode and equals. Sometimes hashCode and equals are just used in unit tests, which is fine, but such a use should probably be called out in a javadoc in the class for clarification. Fixing hashCode and equals for classes that are missing it is probably outside the scope of this PR, but filing an issue with the offending classes (if indeed hashCode and equals are improperly missing) would be helpful.

jon-wei commented:
re: equals() and hashCode(), copying a comment from earlier:

Only BoundDimFilter and SelectorDimFilter have proper equals implementations;

I had the equals() and hashcode() methods for those DimFilters throw an exception when called, and found out that AndDimFilter and OrDimFilter trigger equals() on its sub filters via this call:

    fields.removeAll(Collections.singletonList(null));
But the default equals() implementation is sufficient for this null removal.

I also disabled the proper equals/hashcode methods on BoundDimFilter and SelectorDimFilter, the only problem I saw with doing that in the unit tests was in TimeseriesQueryRunnerTest, where there are some explicit equality checks on SelectorDimFilter:

    Assert.assertEquals(
        Druids.newSelectorDimFilterBuilder()
              .dimension(QueryRunnerTestHelper.marketDimension)
              .value("upfront")
              .build(),
        query.getDimensionsFilter()
    );

jon-wei commented:
@drcrallen Do you have any suggestions for unit tests that you would like to see? 

I added a "filters with extraction functions on null column" test in my commits today, and plan to add some more tests extending the BaseFilterTest introduced by https://github.com/druid-io/druid/pull/2727 when that's merged

jon-wei commented:
I keep seeing travis failures on:

```
ERROR] /home/travis/build/druid-io/druid/processing/src/test/java/io/druid/query/filter/SelectorDimFilterTest.java:[44,43] constructor SelectorDimFilter in class io.druid.query.filter.SelectorDimFilter cannot be applied to given types;
  required: java.lang.String,java.lang.String,io.druid.query.extraction.ExtractionFn
  found: java.lang.String,java.lang.String
  reason: actual and formal argument lists differ in length
```

but I've updated that file:
https://github.com/jon-wei/druid/blob/filter_support/processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java

and this patch passes unit tests on my local machine.

Any ideas on what's happening?

gianm commented:
Try, try again…

drcrallen commented:
@jon-wei there are a lot of changes to utility functions such as hashCode cacheKey toString and equals, but not a lot of coverage for those methods in the classes affected.

jon-wei commented:
@drcrallen Added some tests to SelectoDimFilterTest and BoundDimFilterTests for coverage on those utility methods

jon-wei commented:
re: travis failures, there was a merge conflict in SelectorDimFilterTest that didn't get picked up by github for whatever reason, I rebased to fix it.

fjy commented:
@jon-wei some more merge conflicts

drcrallen commented:
:+1: after travis

jon-wei commented:
Rebased, there are changes to BoundFilter from merging with https://github.com/druid-io/druid/pull/2727
- If extractionFn is defined, the optimization for finding start/end index is not used
- extractionFn is applied in doesMatch() function

jon-wei commented:
rebasing

jon-wei commented:
rebased

jon-wei commented:
rebased, added extractionFn to new InFilter from https://github.com/druid-io/druid/pull/2577 changes

gianm commented on processing/src/test/java/io/druid/query/filter/BoundDimFilterTest.java:
```diff
@@ -68,10 +75,37 @@ public void testSerDesBoundFilter() throws IOException
+    byte[] cacheKey = new byte[] {0x0A, 0x01, 0x00, 0x00, 0x01, (byte) 0xFF, 0x64, 0x69, 0x6D, 0x65, 0x6E, 0x73, 0x69,
+                                  0x6F, 0x6E, (byte) 0xFF, 0x31, 0x35, (byte) 0xFF, 0x31, 0x32, (byte) 0xFF};
+
+    byte[] cacheKeyWithExtract = new byte[] {0x0A, 0x01, 0x00, 0x00, 0x01, (byte) 0xFF, 0x64, 0x69, 0x6D, 0x65, 0x6E, 0x73, 0x69,
```
This test is gonna be really brittle to future changes in how the cache keys are computed. It'd probably be just as good to assert that the cache keys are different for something with extractions and something without extractions.

jon-wei replied:
@gianm changed this to compare cache keys between two filters

gianm commented on processing/src/test/java/io/druid/query/filter/BoundDimFilterTest.java:
```diff
@@ -68,10 +75,37 @@ public void testSerDesBoundFilter() throws IOException
+    BoundDimFilter boundDimFilter = new BoundDimFilter("dimension", "12", "15", null, null, true, null);
+    BoundDimFilter boundDimFilterWithExtract = new BoundDimFilter("dimension", "12", "15", null, null, true, extractionFn);
+
+    int hashCode = 1917945602;
```
This test is also brittle.

jon-wei replied:
@gianm changed this to compare hashcodes from two filters

gianm commented on processing/src/test/java/io/druid/segment/filter/BoundFilterTest.java:
```diff
@@ -77,10 +77,10 @@ public BoundFilterTest(
   public void testLexicographicMatchEverything()
```
Could you add a test somewhere in this file for a filter with extraction functions? Also for SelectorFilterTest and InFilterTest.

jon-wei replied:
@gianm Added JavascriptFilterTest, SearchQueryFilterTest, RegexFilterTest in addition to the current set, added a test for extraction functions to each

gianm commented on processing/src/test/java/io/druid/segment/filter/ExtractionDimFilterTest.java:
```diff
@@ -160,7 +160,7 @@ public ExtractionType getExtractionType()
   public void testEmpty()
   {
-    ExtractionFilter extractionFilter = new ExtractionFilter(
+    SelectorFilter extractionFilter = new SelectorFilter(
```
This should still be an ExtractionFilter because this test is meant to be testing ExtractionFilters. It should keep doing that until they are actually removed.

jon-wei replied:
I'm removing ExtractionFilter but keeping ExtractionDimFilter in this patch

gianm commented on processing/src/test/java/io/druid/segment/filter/InFilterTest.java:
```diff
@@ -138,7 +138,7 @@ public void testMissingColumn()
   private DimFilter toInFilter(String dim, String value, String... values)
   {
-    return new InDimFilter(dim, Lists.asList(value, values));
+    return new InDimFilter(dim, Lists.asList(value, values), null);
```
while you're adding an extractionFn test to InFilterTest, could you please also change this to use the newer filter test style with `assertFilterMatches`? That'll make sure that filtered aggregators work too.

jon-wei replied:
@gianm changed InFilterTest to use assertFilterMatches

gianm commented on docs/content/querying/filters.md:
```diff
@@ -81,13 +86,21 @@ The following matches any dimension values for the dimension `name` between `'ba
+The JavaScript filter supports the use of extraction functions, see [Filtering with Extraction Functions](#filtering-with-extraction-functions) for details.
+
+
+### Extraction filter (Deprecated)
```
Could you remove the (Deprecated) and just leave the note? The reason is that altering the title of a section changes the named anchors and can break incoming links.

jon-wei replied:
@gianm removed Deprecated from title

gianm commented on docs/content/querying/filters.md:
```diff
@@ -81,13 +86,21 @@ The following matches any dimension values for the dimension `name` between `'ba
+`output_value` is equal to the filter `value` and `input_key` is present as dimension.
 
-**Example**
+**Example**		
```
bunch of extraneous spaces here.

jon-wei replied:
@gianm removed spaces

gianm commented on docs/content/querying/filters.md:
```diff
@@ -81,13 +86,21 @@ The following matches any dimension values for the dimension `name` between `'ba
+</div>
 
-Extraction filter matches a dimension using some specific [Extraction function](./dimensionspecs.html#extraction-functions).
+Extraction filter matches a dimension using some specific [Extraction function](./dimensionspecs.html#extractionfunctions).
```
I think the dash in `extraction-functions` was correct.

jon-wei replied:
@gianm restored the dash, think I accidentally removed it..

gianm commented on docs/content/querying/filters.md:
```diff
@@ -133,6 +147,9 @@ Search filters can be used to filter on partial string matches.
 |dimension|The dimension to perform the search over.|yes|
 |query|A JSON object for the type of search. See below for more information.|yes|
```
would also add an `extractionFn` item here (and actually I don't think the block below is needed if you have this; that content would work in this table)

gianm replied:
Hmm actually a lot of other filters don't have these kinds of tables and it'd be nice to have the notice consistent…

Maybe keep the notice and add something here like,

`|extractionFn|[Extraction function](#filtering-with-extraction-functions) to apply to the dimension|no|`

jon-wei replied:
Kept the notices, added an entry for extraction function to the SearchQueryFilter property table

gianm commented on processing/src/main/java/io/druid/query/filter/ExtractionDimFilter.java:
```diff
@@ -111,7 +113,7 @@ public DimFilter optimize()
   public Filter toFilter()
   {
-    return new ExtractionFilter(dimension, value, extractionFn);
+    return new SelectorFilter(dimension, value, extractionFn);
```
IMO you can delete `ExtractionFilter` if you are making this change.

jon-wei replied:
@gianm Deleted ExtractionFilter

gianm commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -51,25 +60,44 @@ public SelectorDimFilter(
+      final List<String> keys = lookup.unapply(this.getValue());
+      final String dimensionName = this.getDimension();
+      if (!keys.isEmpty()) {
+        return new OrDimFilter(Lists.transform(keys, new Function<String, DimFilter>()
```
This could be an InDimFilter (it looks like you copied this from ExtractionDimFilter, which was recently updated to use that)

jon-wei replied:
Updated this to use InDimFilter

gianm commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -51,25 +60,44 @@ public SelectorDimFilter(
+      LookupExtractor lookup = ((LookupExtractionFn) this.getExtractionFn()).getLookup();
+      final List<String> keys = lookup.unapply(this.getValue());
+      final String dimensionName = this.getDimension();
+      if (!keys.isEmpty()) {
```
if `keys.isEmpty()` I wonder why we aren't optimizing this into a filter that selects nothing.

gianm replied:
Actually there might be bigger problems?

This is optimizing `lookup(dim) == val` into `dim IN lookup.unapply(val)` but that doesn't properly take into account the missing value handling. e.g. if retainMissingValues is true, and `lookup("val")` were null, then a literal `"val"` should match `lookup(dim) == val` but it won't match `dim IN lookup.unapply(val)`.

jon-wei replied:
@gianm Added checks within optimize() to return without optimization if:
- retainMissingValues is true
- the selector value and replaceMissingValuesWith are identical

gianm commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -84,6 +112,18 @@ public String getValue()
+  @Override
+  public String toString()
+  {
+    return String.format("%s(%s) = %s", extractionFn, dimension, value);
```
nit: this could be prettier-printed if we just do do `dim = val` when extractionFn is null.

jon-wei replied:
@gianm changed print

jon-wei commented:
Addressing comments from @gianm, also going to add more filter tests to this PR

jon-wei commented:
Addressed comments, rebasing now

This PR also fixes a few issues related to filtering, I've opened issues for them:
https://github.com/druid-io/druid/issues/2772
https://github.com/druid-io/druid/issues/2775
https://github.com/druid-io/druid/issues/2776

jon-wei commented:
rebased

gianm commented on processing/src/main/java/io/druid/query/filter/ExtractionDimFilter.java:
```diff
@@ -98,11 +102,25 @@ public DimFilter optimize()
   {
     if (this.getExtractionFn() instanceof LookupExtractionFn
```
why not `return new SelectorDimFilter(dimension, value, extractionFn).optimize();`?

jon-wei replied:
@gianm good idea, I'll change that

jon-wei replied:
@gianm changed to use SelectorDimFilter optimize()

gianm commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -51,25 +62,51 @@ public SelectorDimFilter(
+      // If retainMissingValues is true, we can't do an unapply()-based optimization, there may be
+      // row values that match the selector value but are not included in the lookup map.
+      if (exFn.isRetainMissingValue()) {
+        return this;
```
I think we actually could optimize this. `lookup(dim) = val` with retainMissingValue will match if,

- dimValue is in lookup.unapply(value)
- or, lookup.apply(dimValue) is null and dimValue == value

jon-wei replied:
@gianm updated optimization for SelectorDimFilter, it will now add the selector value to the InDimFilter when optimizing with retainMissingValues (or return a single SelectorFilter if the reverse lookup was empty)

gianm commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -38,9 +39,14 @@
+  private final ExtractionFn extractionFn;
 
   @JsonCreator
-  public InDimFilter(@JsonProperty("dimension") String dimension, @JsonProperty("values") List<String> values)
```
This could be optimized in a similar way to the SelectorDimFilter.

jon-wei replied:
@gianm Can we handle the InDimFilter optimization in a separate PR?

gianm replied:
sure, you can file an issue for anything you don't think is appropriate for this PR.

gianm commented on processing/src/main/java/io/druid/segment/filter/DimensionPredicateFilter.java:
```diff
@@ -37,27 +38,54 @@
       return selector.getBitmapFactory().makeEmptyImmutableBitmap();
     }
+    Indexed<String> dimValues = selector.getDimensionValues(dimension);
+    if (dimValues == null || dimValues.size() == 0) {
```
I don't think `dimValues.size() == 0` is correct; a single value doesn't HAVE to be null. This should also check `dimValues.get(0) == null`

gianm replied:
nevermind I was reading this as `size() == 1`

gianm commented on processing/src/main/java/io/druid/segment/filter/JavaScriptFilter.java:
```diff
@@ -52,8 +55,11 @@ public ImmutableBitmap getBitmapIndex(final BitmapIndexSelector selector)
       final Indexed<String> dimValues = selector.getDimensionValues(dimension);
       ImmutableBitmap bitmap;
-      if (dimValues == null) {
+      if (dimValues == null || dimValues.size() == 0) {
```
Similar comment to the other one on `dimValues.size() == 0`

gianm replied:
nevermind, I was reading this as `size() == 1`

gianm commented on processing/src/main/java/io/druid/segment/filter/SelectorFilter.java:
```diff
@@ -31,25 +40,68 @@
   {
     this.dimension = dimension;
-    this.value = value;
+    this.value = Strings.nullToEmpty(value);;
```
double `;;`

jon-wei replied:
@gianm removed extra semicolon

gianm commented on processing/src/main/java/io/druid/segment/filter/SelectorFilter.java:
```diff
@@ -31,25 +40,68 @@
+    }
+  }
+
+  private List<Filter> makeFilters(BitmapIndexSelector selector)
```
`makeFiltersUsingExtractionFn`? this method does not make sense to call if extractionFn is null, IMO that should be clearer.

jon-wei replied:
@gianm renamed function

gianm commented on processing/src/test/java/io/druid/segment/filter/SelectorFilterTest.java:
```diff
@@ -81,54 +86,107 @@ public SelectorFilterTest(
+    LookupExtractionFn lookupFn3 = new LookupExtractionFn(mapExtractor3, false, null, false, true);
+    assertFilterMatches(new SelectorDimFilter("dim0", null, lookupFn3), ImmutableList.of("0", "1", "2", "3", "4", "5"));
+
+    // optimize() tests, these cases cannot be optimized and the filter should be unchanged
```
BaseFilterTest tests all filters with and without optimizing, assuming you're using its `select*` methods (which you are). So there isn't any reason to explicitly include optimization tests.

gianm replied:
(unless you want to test that the optimized form is a certain specific thing – but you aren't doing that)

jon-wei replied:
I made those tests before the BaseFilterTest change that added the optimization parameter, I'll change these optimize() specific tests to check that they return a converted filter of the proper form in addition to results

jon-wei replied:
updated tests

gianm commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -51,25 +63,57 @@ public SelectorDimFilter(
+        // If retainMissingValues is true, there may be row values that match the selector value
+        // but are not included in the lookup map. Match on the selector value as well.
+        if (exFn.isRetainMissingValue()) {
+          keys.add(value);
```
I believe this is only safe to do if `value` is not actually in the lookup (if it is then `value` would map to something else and should not match this filter).

gianm commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -51,25 +63,57 @@ public SelectorDimFilter(
+      } else {
+        // Nothing in the lookup would match, just use a plain selector filter.
+        if (exFn.isRetainMissingValue()) {
+          return new SelectorDimFilter(dimension, value, null);
```
similar comment as above

gianm replied:
I think this would be clearer if the `else` and `if` were combined into an `else if`. It made me do a double-take originally.

Would be even clearer to include an `else { return this }` too, just to cover all the cases explicitly.

jon-wei replied:
@gianm 

I rewrote the optimize() function, it takes into account this comment and the one regarding checking for the selector value in the lookup map.

Also added tests for that behavior

gianm commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -51,25 +63,61 @@ public SelectorDimFilter(
+      // in the lookup map. Match on the selector value as well.
+      // If the selector value is overwritten in the lookup map, don't add selector value to keys.
+      if (exFn.isRetainMissingValue() && mappingForValue == null) {
+        keys.add(convertedValue);
```
`keys` isn't guaranteed to be mutable, so it'd be better to copy it and add `convertedValue` to the copy.

gianm replied:
this looks good to me other than that

jon-wei replied:
done

gianm commented:
lgtm 👍
